source/common/http/filter/gzip_filter.cc

```diff
@@ -0,0 +1,274 @@
+  if (!end_stream && !skip_compression_ && isMinimumContentLength(headers) &&
+      isContentTypeAllowed(headers) && isCacheControlAllowed(headers) &&
+      isLastModifiedAllowed(headers) && isTransferEncodingAllowed(headers) &&
+      !headers.ContentEncoding() && isEtagAllowed(headers)) {
+
+    insertVaryHeader(headers);
+    headers.removeContentLength();
+    headers.insertContentEncoding().value(Http::Headers::get().ContentEncodingValues.Gzip);
+    compressor_.init(config_->compressionLevel(), config_->compressionStrategy(),
+                     config_->windowBits(), config_->memoryLevel());
+
+  } else {
+    skip_compression_ = true;
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+FilterDataStatus GzipFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (skip_compression_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  const uint64_t n_data{data.length()};
+
+  if (n_data) {
+    compressor_.compress(data, compressed_data_);
+  }
+
+  if (end_stream) {
+    compressor_.flush(compressed_data_);
+  }
+
+  if (compressed_data_.length()) {
+    data.drain(n_data);
+    data.move(compressed_data_);
+    return Http::FilterDataStatus::Continue;
+  }
+
+  return Http::FilterDataStatus::StopIterationNoBuffer;
+}
+
+bool GzipFilter::isAcceptEncodingAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* accept_encoding = headers.AcceptEncoding();
+  // TODO(gsagula): Since gzip is the only available content-encoding in Envoy at the moment,
+  // order/priority of preferred server encodings is disregarded (RFC2616-14.3). Replace this
+  // with a data structure that parses Accept-Encoding values and allows fast lookup of
+  // key/priority. Also, this should be part of some utility library.
+  if (accept_encoding) {
+    bool is_identity{true};  // true if does not exist or is followed by `q=0`.
+    bool is_wildcard{false}; // true if found and not followed by `q=0`.
+    bool is_gzip{false};     // true if exists and not followed by `q=0`.
+
+    for (const auto token :
+         StringUtil::splitToken(headers.AcceptEncoding()->value().c_str(), ",", true)) {
+      const auto values = StringUtil::splitToken(token, ";");
+      const auto value = StringUtil::trim(values.at(0));
+      const auto q_value =
+          values.size() > 1 ? StringUtil::trim(values.at(1)) : absl::string_view{nullptr, 0};
+
+      if (value == Http::Headers::get().AcceptEncodingValues.Gzip) {
+        is_gzip = q_value.empty() ? true : (q_value != ZeroQvalueString);
+      }
+      if (value == Http::Headers::get().AcceptEncodingValues.Identity) {
+        is_identity = q_value.empty() ? false : (q_value == ZeroQvalueString);
+      }
+      if (value == Http::Headers::get().AcceptEncodingValues.Wildcard) {
+        is_wildcard = q_value.empty() ? true : (q_value != ZeroQvalueString);
+      }
+    }
+
+    return ((is_gzip && is_identity) || is_wildcard);
```

Comment:
The semantics here look subtle.  Give an HTTP spec reference for this in a comment.